src/smt/proof_final_callback.cpp

```diff
@@ -133,6 +136,47 @@ bool ProofFinalCallback::shouldUpdate(std::shared_ptr<ProofNode> pn,
                              << " ---> " << eq[1] << std::endl;
     }
   }
+  if (options().proof.checkProofSteps)
+  {
+    Node conc = pn->getResult();
+    ProofChecker* pc = pnm->getChecker();
+    if (pc->getPedanticLevel(r) == 0)
+    {
+      // no need to check
+    }
+    else
+    {
+      std::vector<Node> premises;
+      const std::vector<std::shared_ptr<ProofNode>>& pnc = pn->getChildren();
+      for (const std::shared_ptr<ProofNode>& pncc : pnc)
+      {
+        premises.push_back(pncc->getResult());
+      }
+      NodeManager* nm = NodeManager::currentNM();
+      Node query = nm->mkNode(IMPLIES, nm->mkAnd(premises), conc);
+      // trust the rewriter here, since the subsolver will rewrite anyways
+      query = rewrite(query);
+      // We use the original form of the query, which is a logically
+      // stronger formula. This may make it possible or easier to prove.
+      query = SkolemManager::getOriginalForm(query);
```

Comment:
does this leads to witness terms in the input (for skolems that are not purification ones)? I don't remember now. And in that case, we are fine for handling that?

Reply:
We do not use witness terms anymore in original forms.  This indeed would lead to issues here, but this not the case.

Reply:
I see, cool!